src/tree.jl

```diff
@@ -249,3 +313,125 @@ function fill_unpopulated!(tree::SARSOPTree, b_idx::Int)
+            tree.bm.bin_levels_nodes[level_i][b_idx][:prev_error] = err * err
+            tree.bm.bin_levels[level_i][:bin_error][key] = err * err
+            
+            value = lb_val
+            tree.bm.bin_levels[level_i][:bin_count] = Dict(key => bin_count + 1)
+        end
+        tree.bm.bin_levels[level_i][:bin_value][key] = value
+    end
+    tree.bm.previous_lowerbound[b_idx] = lb_val
+end
+
+function update_bin_node!(tree::SARSOPTree, b_idx::Int)
+    lb_val = tree.V_lower[b_idx]
+    up_val = tree.V_upper[b_idx]
+   
+    if !haskey(tree.bm.bin_levels_nodes[1], b_idx)
+        return initialize_bin_node!(tree, b_idx)
+    end
+    
+    for level_i in 1:tree.bm.num_levels
+        key = tree.bm.bin_levels_nodes[level_i][b_idx][:key]
+        
+        bin_count = get(tree.bm.bin_levels[level_i][:bin_count], key, 0)
+        if bin_count == 1
+            err = up_val - lb_val
+            tree.bm.bin_levels_nodes[level_i][b_idx][:prev_error] = err * err
+            tree.bm.bin_levels[level_i][:bin_error][key] = err * err
+        else
+            err = tree.bm.bin_levels[level_i][:bin_value][key] - lb_val
+            tree.bm.bin_levels[level_i][:bin_error][key] -= tree.bm.bin_levels_nodes[level_i][b_idx][:prev_error]
+            tree.bm.bin_levels_nodes[level_i][b_idx][:prev_error] = err * err
+            tree.bm.bin_levels[level_i][:bin_error][key] += err * err
+        end
+
+        tree.bm.bin_levels[level_i][:bin_value][key] = (tree.bm.bin_levels[level_i][:bin_value][key] * bin_count + lb_val - tree.bm.previous_lowerbound[b_idx]) / bin_count
+    end
+    tree.bm.previous_lowerbound[b_idx] = lb_val
+end
+
+function get_bin_value(tree::SARSOPTree, b_idx::Int)
+    
+    lb_val = tree.V_lower[b_idx]
+    ub_val = tree.V_upper[b_idx]
+    
+    key = tree.bm.bin_levels_nodes[1][b_idx][:key]
+    if tree.bm.bin_levels[1][:bin_count][key] == 1
+        return ub_val
+    else
+        smallest_error = Inf
+        best_level = 0
+        best_key = key
+        for level_i in 1:tree.bm.num_levels
+            key = tree.bm.bin_levels_nodes[level_i][b_idx][:key]
+            if tree.bm.bin_levels[level_i][:bin_error][key] + 1e-10 < smallest_error
+                best_level = level_i
+                smallest_error = tree.bm.bin_levels[level_i][:bin_error][key]
+                best_key = key
+            end
+        end
+        
+        if tree.bm.bin_levels[best_level][:bin_value][best_key] > ub_val + 1e-10
+            return ub_val
+        elseif tree.bm.bin_levels[best_level][:bin_value][best_key] + 1e-10 < lb_val
+            return lb_val
+        else    
+            return tree.bm.bin_levels[best_level][:bin_value][best_key]
+        end
+    end
+end
+
+function max_entropy(n::Int)
+    return -1 * ((1.0 / n) * log(1.0 / n)) * n
+end
+
+function entropy(b::AbstractVector)
+    ent = 0.0
+    for b_i in b
+        b_i > 0 && (ent -= b_i * log(b_i))
+    end
+    return ent
+end
```

Comment:
This works well enough for dense vectors, but could you also provide a method optimized for sparse vectors?

Our beliefs are all sparse vectors, so I think this to be necessary.

Reply:
Why are beliefs sparse vectors? Aren't beliefs often dense?

Reply:
Often? I'm not sure. If there's any sort of mixed observability the beliefs become quite sparse. We used rock sample as a benchmark when creating this package where the beliefs tend to be very sparse.

Reply:
I guess that's true. Seems like there are often dense "regions" though. Someday, it would be interesting to see if we could leverage this with different sets of alpha vectors for different regions. I wonder if that is essentially what CPP SARSOP does for MOMDPs. (not asking for an answer here, just musing)

Reply:
Added in [4c5b380](https://github.com/JuliaPOMDP/NativeSARSOP.jl/pull/18/commits/4c5b3806408b1d3de71f5fd1f1691043625f6992)